File: Managed Source/Arsenal.ImageMounter/IO/BufferedBinaryWriter.vb
Imports System.Diagnostics.CodeAnalysis
Imports System.IO
Imports System.Text

Namespace IO

    ''' <summary>
    ''' Buffered version of the BinaryWriter class. Writes to a MemoryStream internally and flushes
    ''' writes out contents of MemoryStream when WriteTo() or ToArray() are called.
    ''' </summary>
    Public Class BufferedBinaryWriter
        Inherits BinaryWriter

        ''' <summary>
        ''' Creates a new instance of BufferedBinaryWriter.
        ''' </summary>
        ''' <param name="encoding">Specifies which text encoding to use.</param>
        Public Sub New(encoding As Encoding)
            MyBase.New(New MemoryStream, encoding)
        End Sub

        ''' <summary>
        ''' Creates a new instance of BufferedBinaryWriter using System.Text.Encoding.Unicode text encoding.
        ''' </summary>
        Public Sub New()
            MyBase.New(New MemoryStream, Encoding.Unicode)
        End Sub

        ''' <summary>
        ''' Writes current contents of internal MemoryStream to another stream and resets
        ''' this BufferedBinaryWriter to empty state.
        ''' </summary>
        ''' <param name="stream"></param>
        Public Sub WriteTo(stream As Stream)
            Flush()
            With DirectCast(BaseStream, MemoryStream)
                .WriteTo(stream)
                .SetLength(0)
                .Position = 0
            End With
            stream.Flush()
        End Sub

        ''' <summary>
        ''' Extracts current contents of internal MemoryStream to a new byte array and resets
        ''' this BufferedBinaryWriter to empty state.
        ''' </summary>
        Public Function ToArray() As Byte()
            Flush()
            With DirectCast(BaseStream, MemoryStream)
                ToArray = .ToArray()
                .SetLength(0)
                .Position = 0
            End With
        End Function

        ''' <summary>
        ''' Clears contents of internal MemoryStream.
        ''' </summary>
        Public Sub Clear()
            If _IsDisposed = True Then
                Return
            End If

            With DirectCast(BaseStream, MemoryStream)
                .SetLength(0)
                .Position = 0
            End With
        End Sub

        Public ReadOnly Property IsDisposed As Boolean

        Protected Overrides Sub Dispose(disposing As Boolean)
            _IsDisposed = True

            MyBase.Dispose(disposing)
        End Sub

    End Class

End Namespace

File: Managed Source/Arsenal.ImageMounter/IO/Streams/BufferedBinaryWriter.cs
//  
//  Copyright (c) 2012-2024, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
//  This source code and API are available under the terms of the Affero General Public
//  License v3.
// 
//  Please see LICENSE.txt for full license terms, including the availability of
//  proprietary exceptions.
//  Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
// 

using Arsenal.ImageMounter.Extensions;
using LTRData.Extensions.Buffers;
using System;
using System.IO;
using System.Text;



namespace Arsenal.ImageMounter.IO.Streams;

/// <summary>
/// Buffered version of the BinaryWriter class. Writes to a MemoryStream internally and flushes
/// writes out contents of MemoryStream when WriteTo() or ToArray() are called.
/// </summary>
public class BufferedBinaryWriter : BinaryWriter
{

    /// <summary>
    /// Creates a new instance of BufferedBinaryWriter.
    /// </summary>
    /// <param name="encoding">Specifies which text encoding to use.</param>
    public BufferedBinaryWriter(Encoding encoding)
        : base(new MemoryStream(), encoding)
    {
    }

    /// <summary>
    /// Creates a new instance of BufferedBinaryWriter using System.Text.Encoding.Unicode text encoding.
    /// </summary>
    public BufferedBinaryWriter()
        : base(new MemoryStream(), Encoding.Unicode)
    {
    }

    /// <summary>
    /// Writes current contents of internal MemoryStream to another stream and resets
    /// this BufferedBinaryWriter to empty state.
    /// </summary>
    /// <param name="stream"></param>
    public void WriteTo(Stream stream)
    {
        Flush();

        var baseStream = (MemoryStream)BaseStream;
        baseStream.WriteTo(stream);
        baseStream.SetLength(0);
        baseStream.Position = 0;

        stream.Flush();
    }

    /// <summary>
    /// Extracts current contents of internal MemoryStream to a new byte array and resets
    /// this BufferedBinaryWriter to empty state.
    /// </summary>
    public byte[] ToArray()
    {
        Flush();

        var ms = (MemoryStream)BaseStream;
        var ToArrayRet = ms.ToArray();
        ms.SetLength(0);
        ms.Position = 0;

        return ToArrayRet;
    }

    /// <summary>
    /// Provides direct access to the byte buffer used by this instance.
    /// Only valid until next Write/Clear/ToArray etc modifying methods.
    /// </summary>
    /// <returns>Span access to byte buffer</returns>
    public Span<byte> AsSpan() => ((MemoryStream)BaseStream).AsSpan();

    /// <summary>
    /// Clears contents of internal MemoryStream.
    /// </summary>
    public void Clear()
    {
        if (IsDisposed == true)
        {
            return;
        }

        var ms = (MemoryStream)BaseStream;
        ms.SetLength(0);
        ms.Position = 0;
    }

    public bool IsDisposed { get; private set; }

    protected override void Dispose(bool disposing)
    {
        IsDisposed = true;

        base.Dispose(disposing);
    }
}